remappers/hostmetrics/disk.go

```diff
@@ -0,0 +1,129 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package hostmetrics
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"golang.org/x/exp/constraints"
+)
+
+func remapDiskMetrics(src, out pmetric.MetricSlice, _ pcommon.Resource, dataset string) error {
+	for i := 0; i < src.Len(); i++ {
+		metric := src.At(i)
+		switch metric.Name() {
+		case "system.disk.io", "system.disk.operations":
+			dataPoints := metric.Sum().DataPoints()
+			for j := 0; j < dataPoints.Len(); j++ {
+				dp := dataPoints.At(j)
+
+				device, ok := dp.Attributes().Get("device")
+				if !ok {
+					continue
+				}
+
+				var multiplier int64 = 1
+				if direction, ok := dp.Attributes().Get("direction"); ok {
+					name := metric.Name()
+					timestamp := dp.Timestamp()
+					value := dp.IntValue()
+					addDiskMetric(out, dataset, name, device.Str(), direction.Str(), timestamp, value, multiplier)
+				}
+			}
+		case "system.disk.operation_time", "system.disk.io_time":
+			var multiplier float64
+			dataPoints := metric.Sum().DataPoints()
+			for j := 0; j < dataPoints.Len(); j++ {
+				dp := dataPoints.At(j)
+				timestamp := dp.Timestamp()
+				value := dp.DoubleValue()
+				multiplier = 1000 // Elastic saves this value in milliseconds
+
+				device, ok := dp.Attributes().Get("device")
+				if !ok {
+					continue
+				}
+
+				if direction, ok := dp.Attributes().Get("direction"); ok {
+					addDiskMetric(out, dataset, metric.Name(), device.Str(), direction.Str(), timestamp, value, multiplier)
+				} else {
+					addDiskMetric(out, dataset, metric.Name(), device.Str(), "", timestamp, value, multiplier)
+				}
+			}
+		case "system.disk.pending_operations":
+			dataPoints := metric.Sum().DataPoints()
+			for j := 0; j < dataPoints.Len(); j++ {
+				dp := dataPoints.At(j)
+				timestamp := dp.Timestamp()
+				value := dp.IntValue()
+
+				device, ok := dp.Attributes().Get("device")
+				if !ok {
+					continue
+				}
+
+				var multiplier int64 = 1
+				addDiskMetric(out, dataset, metric.Name(), device.Str(), "", timestamp, value, multiplier)
+			}
+		}
+	}
+	return nil
+}
+
+func addDiskMetric[T interface {
+	constraints.Integer | constraints.Float
+}](out pmetric.MetricSlice,
+	dataset, name, device, direction string,
+	timestamp pcommon.Timestamp,
+	value, multiplier T) {
+
+	metricsToAdd := map[string]string{
```

Comment:
Same here I would define this mapping on top as constant.

Reply:
+1 on what @ChrsMark said. I would prefer if `metricsToAdd` is outside `addDiskMetric` as a global private var. In the current logic `metricsToAdd` might escape to heap leading to more GC work.